### PR TITLE
Add YOLOv8 for Camera Object Detection Node

### PR DESCRIPTION
**Pull request by Gongsta**
Clean fork of justin's YOLOv8 PR https://github.com/WATonomous/wato_monorepo/pull/34/

### CODE REVIEW

**Comment by Gongsta** [comment]
Tested and visualized on Foxglove 

Screenshot 2024-01-25 at 6 31 37 PM

File: src/wato_msgs/common_msgs/msg/BoundingBox.msg
[the remark this answers is not in the document]
Comment (Edwardius) [reply src/wato_msgs/common_msgs/msg/BoundingBox.msg:1]: also, any perception messages should go under `perception_msgs`

**Review by Edwardius — CHANGES_REQUESTED** [review]
Looks great, alot of the comments are about handling multiple cameras, which I feel should be handled sooner rather than later. We got a lot of ground to cover to build our AV as fast as we can, that's why I feel that we should make this program handle it.

resolve whenever u can

File: src/perception/camera_detection/camera_detection/yolov8_detection.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+        self.declare_parameter('camera_topic', '/camera/right/image_color')
+        self.declare_parameter('publish_vis_topic', '/annotated_img')
+        self.declare_parameter('publish_obstacle_topic', '/obstacles')
+        self.declare_parameter('model_path', '/perception_models/yolov8s.pt')
+        self.declare_parameter('image_size', 480)
+        self.declare_parameter('compressed', False)
+        
+        self.camera_topic = self.get_parameter('camera_topic').value
+        self.publish_vis_topic = self.get_parameter('publish_vis_topic').value
```
[the remark this answers is not in the document]
Comment (Edwardius) [reply src/perception/camera_detection/camera_detection/yolov8_detection.py:39]: OH! actually better yet you can map a publisher and subscriber to a specific camera topic

eg. self.publishers['/camera/left'] = <publisher>

File: src/perception/camera_detection/camera_detection/yolov8_detection.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        annotator_img = annotator.result()
+        return (processed_detections, annotator_img)
+
+    def publish_vis(self, annotated_img, feed):
+        # Publish visualizations
+        imgmsg = self.cv_bridge.cv2_to_imgmsg(annotated_img, "bgr8")
+        imgmsg.header.frame_id = "camera_{}_link".format(feed)
+        self.vis_publisher.publish(imgmsg)
+
+        # TODO: Support reading stereo (left and right images)
```
[the remark this answers is not in the document]
Comment (Edwardius) [reply src/perception/camera_detection/camera_detection/yolov8_detection.py:147]: Ignore this suggestion, should maybe consider: https://github.com/WATonomous/wato_monorepo/pull/48#discussion_r1468919888

File: src/wato_msgs/common_msgs/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+# Search for dependencies required for building this package
+find_package(ament_cmake REQUIRED) # ROS2 build tool
+find_package(std_msgs REQUIRED) # ROS2 common messages
+find_package(geometry_msgs REQUIRED)
+find_package(sensor_msgs REQUIRED)
+find_package(builtin_interfaces REQUIRED)
+find_package(rosidl_default_generators REQUIRED)
+
+set(msg_files
```
[the remark this answers is not in the document]
Comment (Gongsta) [reply src/wato_msgs/common_msgs/CMakeLists.txt:21]: Agreed, removing all these in favor of standard vision_msgs

File: docker/perception/camera_object_detection/camera_object_detection.Dockerfile [hunk 4]
```diff
@@ -21,12 +22,22 @@ FROM ${BASE_IMAGE} as dependencies
...
 # Install Rosdep requirements
 COPY --from=source /tmp/colcon_install_list /tmp/colcon_install_list
-RUN apt-fast install -qq -y --no-install-recommends $(cat /tmp/colcon_install_list)
+RUN apt-fast update --fix-missing && apt-fast install -qq -y --no-install-recommends $(cat /tmp/colcon_install_list)
 
 # Copy in source code from source stage
 WORKDIR ${AMENT_WS}
 COPY --from=source ${AMENT_WS}/src src
 
+# Install pip
```
[the remark this answers is not in the document]
Comment (Gongsta) [reply docker/perception/camera_object_detection/camera_object_detection.Dockerfile:31]: ack

File: docker/perception/camera_object_detection/camera_object_detection.Dockerfile [hunk 5]
```diff
@@ -41,6 +52,11 @@ RUN . /opt/ros/$ROS_DISTRO/setup.sh && \
     colcon build \
         --cmake-args -DCMAKE_BUILD_TYPE=Release
 
+
+# Download yolov8 model
+RUN sudo mkdir -m 777 -p /perception_models
+RUN wget -P /perception_models https://github.com/ultralytics/assets/releases/download/v0.0.0/yolov8s.pt
```
[the remark this answers is not in the document]
Comment (Gongsta) [reply docker/perception/camera_object_detection/camera_object_detection.Dockerfile:58]: done

File: src/perception/camera_detection/camera_detection/yolov8_detection.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+        # fill header for obstacle list
+        obstacle_list.header.frame_id = msg.header.frame_id
+
+        # populate obstacle list
+        if detections is not None and len(detections):
+            for detection in detections:
+                obstacle = Obstacle()
+
+                # fill header for obstacle
+                obstacle.header.frame_id = msg.header.frame_id
```
[the remark this answers is not in the document]
Comment (Gongsta) [reply src/perception/camera_detection/camera_detection/yolov8_detection.py:167]: yup

File: docker/perception/camera_object_detection/camera_object_detection.Dockerfile [hunk 7]
```diff
@@ -21,12 +22,22 @@ FROM ${BASE_IMAGE} as dependencies
 
 # Install Rosdep requirements
 COPY --from=source /tmp/colcon_install_list /tmp/colcon_install_list
-RUN apt-fast install -qq -y --no-install-recommends $(cat /tmp/colcon_install_list)
+RUN apt-fast update --fix-missing && apt-fast install -qq -y --no-install-recommends $(cat /tmp/colcon_install_list)
```
[the remark this answers is not in the document]
Comment (Gongsta) [reply docker/perception/camera_object_detection/camera_object_detection.Dockerfile:25]: was due to storage, this was bad fix

File: docker/perception/camera_object_detection/camera_object_detection.Dockerfile [hunk 8]
```diff
@@ -27,6 +32,11 @@ RUN apt-fast install -qq -y --no-install-recommends $(cat /tmp/colcon_install_li
 WORKDIR ${AMENT_WS}
 COPY --from=source ${AMENT_WS}/src src
 
+# Install python packages
+COPY src/perception/camera_object_detection/requirements.txt requirements.txt
+RUN python3 -m pip install -r requirements.txt
+RUN rm requirements.txt
+
```
Comment (Edwardius) [inline docker/perception/camera_object_detection/camera_object_detection.Dockerfile:39]: move to line 26 so it doesn't have to reinstall stuff after a file change

File: modules/docker-compose.perception.yaml [hunk 9]
```diff
@@ -24,8 +24,17 @@ services:
...
+    deploy:
+      resources:
+        reservations:
+          devices:
+            - driver: nvidia
+              count: 1
+              capabilities: [ gpu ]
+    command: /bin/bash -c "ros2 launch camera_object_detection nuscenes_launch.py"
+    volumes:
+      - /mnt/wato-drive2/perception_models/:/perception_models
```
Comment (Edwardius) [inline modules/docker-compose.perception.yaml:37]: plz mount only the weight file, not the entire weights folder

**Review by Edwardius — CHANGES_REQUESTED** [review]
Please add test: Spawn node, send a single dummy image message, ensure that the node outputs bboxes

**Review by Edwardius — APPROVED** [review]
LGTM: TODO for later, handle multi-camera with batch inference.

**Pull request merged**